DataStore2/init.lua

```diff
@@ -544,42 +543,58 @@ function DataStore2.__call(_, dataStoreName, player)
 
 	setmetatable(dataStore, DataStoreMetatable)
 
-	local event, fired = Instance.new("BindableEvent"), false
+	local saveFinishedEvent, isSaveFinished = Instance.new("BindableEvent"), false
+	local bindToCloseEvent = Instance.new("BindableEvent")
 
-	game:BindToClose(function()
-		if not fired then
-			spawn(function()
-				player.Parent = nil -- Forces AncestryChanged to fire and save the data
-			end)
+	local bindToCloseCallback = function()
+		if isSaveFinished == false then
+			bindToCloseEvent:Fire()
 
-			event.Event:Wait()
+			saveFinishedEvent.Event:Wait()
 		end
 
 		local value = dataStore:Get(nil, true)
 
 		for _, bindToClose in ipairs(dataStore.bindToClose) do
 			bindToClose(player, value)
 		end
+	end
+
+	local success, errorMessage = pcall(function()
+		game:BindToClose(function()
+			if bindToCloseCallback == nil then
+				return
+			end
+	
+			bindToCloseCallback()
+		end)
 	end)
+	if success == false then
+		warn(errorMessage)
+	end
 
-	local playerLeavingConnection
-	playerLeavingConnection = player.AncestryChanged:Connect(function()
-		if player:IsDescendantOf(game) then return end
-		playerLeavingConnection:Disconnect()
+	Promise.race({
+		Promise.fromEvent(bindToCloseEvent.Event),
+		Promise.fromEvent(player.AncestryChanged, function()
+			return player:IsDescendantOf(game) == false
+		end),
+	}):andThen(function()
 		dataStore:SaveAsync():andThen(function()
 			print("player left, saved", dataStoreName)
 		end):catch(function(error)
 			-- TODO: Something more elegant
 			warn("error when player left!", error)
 		end):finally(function()
-			event:Fire()
-			fired = true
+			isSaveFinished = true
+			saveFinishedEvent:Fire()
 		end)
 
-		delay(40, function() --Give a long delay for people who haven't figured out the cache :^(
+		--Give a long delay for people who haven't figured out the cache :^(
+		return Promise.delay(40):andThen(function() 
 			DataStoreCache[player] = nil
+			bindToCloseCallback = nil
 		end)
-	end)
+	end):catch(warn)
```

Comment:
Wrap this in an anonymous function that warns and gives the source, like the other `warn`.

Reply:
I added more info to the warning.  I wasn't quite sure what to tell the user though since this promise chain should never error.  `Promise.fromEvent` can only resolve as far as I know, so the `Promise.race` should never error.  Same with `Promise.delay`.

Reply:
Then you shouldn't need this at all, and we would rather this fatally error since something has gone terribly wrong, right?

Reply:
Good point.  Removed in commit 91ade9e.